openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,323 @@
+)
+
+var _ = initializer.WantsExternalKubeInformerFactory(&managementCPUsOverride{})
+var _ = admission.MutationInterface(&managementCPUsOverride{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &managementCPUsOverride{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+// managementCPUsOverride presents admission plugin that will replace pod containers CPUs requests to a new management
+// resource
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister   corev1listers.NamespaceLister
+	nodeLister corev1listers.NodeLister
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	a.nodeLister = kubeInformers.Core().V1().Nodes().Lister()
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	if a.nodeLister == nil {
+		return fmt.Errorf("%s did not get a node lister", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving namespace %q: %v", PluginName, attr.GetNamespace(), err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isNamespaceAllowed(ns.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+		stripWorkloadAnnotations(pod.Annotations)
+
+		return nil
+	}
+
+	if !isPodManaged(pod.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+
+		return nil
+	}
+
+	nodes, err := a.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving nodes: %v", PluginName, err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isManagementResourceAvailable(nodes) {
+		klog.Warningf("%s no nodes with resource %s is available", PluginName, managedCapacityLabel)
+
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+		stripWorkloadAnnotations(pod.Annotations)
+		return nil
+	}
```

Comment:
this block is super odd. We check all nodes whether there is any with the management resource available potentially prior to scheduling? Why?

Reply:
Looks like a hack for something that should be solved elsewhere

Reply:
See enhancement PR, https://github.com/openshift/enhancements/pull/703/files#diff-bdf426b496fe2c7122db9eeaf303b30ff4561b4f7ac157c21996edff1e67e239R234. This feature is cluster wide, so all nodes must have the resources defined in order to activate it.

Reply:
If we don't wait for all nodes to advertise the resource, then we could end up in a situation where in the middle of enabling the feature part of the cluster is unschedulable and either won't form or falls apart. Until we can come up with a better approach, we decided to use a "one way switch" that waits for all nodes to advertise the resource before turning on partitioning and then keep it on after that point, even if a new node appears.

Where are we persisting that state?

Reply:
it possible that a user will add a new node to the cluster, so I prefer to not have a persisting state, also access to the cache should be relatively cheap

Reply:
Dealing with the case where a user adds a new node is exactly what the persistent state is supposed to protect us from. In that case, we still want to mutate the pod definitions, even if it means that node never becomes ready because important pods can't be scheduled on it. The alternative is to have the node mis-configured with pods that are not partitioned, and there would be no indication of that error. At least if the node is broken, there will be an indication. See https://github.com/openshift/enhancements/pull/703/files#diff-bdf426b496fe2c7122db9eeaf303b30ff4561b4f7ac157c21996edff1e67e239R641-R647

@smarterclayton @mrunalp please correct me if I got any of that reasoning wrong.

Reply:
I read the specific section, and I hardly can imagine the situation when the node will not have the partition resource, at least for now we explicitly set the management resource under the kubelet during the node status update. (@rphillips Please correct me if I wrong).
Regarding the state, how can I know if it the final state for the cluster, maybe one of the master nodes started and our admission plugin started on it, but all other nodes still weren't added?

Reply:
It still unclear how to save the state of nodes, we can not really update any objects via API under the admission plugin.

Reply:
We may have to deal with the state question in the next release. When we have an API to enable this feature, that would give us a place for a controller to manage the on/off state based on observing nodes. Then this hook could look at the API resource to determine whether it should partition each type of workload. Maybe you can include a comment in the code explaining that the feature gap only applies to multi-node clusters.